server/src/main/java/io/crate/planner/operators/JoinPlanBuilder.java

```diff
@@ -329,12 +329,18 @@ private static Symbol removeParts(Map<Set<RelationName>, Symbol> queryParts, Rel
 
     @Nullable
     private static <V> V removeMatch(Map<Set<RelationName>, V> valuesByNames, Set<RelationName> names, RelationName nextName) {
+        V match = null;
         for (RelationName name : names) {
             V v = valuesByNames.remove(Set.of(name, nextName));
             if (v != null) {
-                return v;
+                // 'names' set is a pair taken from the beginning of orderedRelationNames.
+                // We get a first match out of (names[0], nextName), (names[1], nextName) to follow the best order
+                // but we keep iterating to remove unused pairs as one table can be part of multiple join conditions.
+                // For query SELECT * from t1 join t2 ON t1.x = t2.x join t3 on (t1.x = t3.x and t2.x = t3.x)
+                // names is (t1, t2) and nextName is t3. We pick up (t1, t3) and remove unused (t2, t3).
+                match = v;
```

Comment:
This is added since now we pick up all pairs.  (`t1,t3)` pair was never created before